formidable/forms/__init__.py

```diff
@@ -48,10 +48,45 @@ def __init__(self, *args, **kwargs):
         super(BaseDynamicForm, self).__init__(*args, **kwargs)
         self._bound_fields_cache = FormidableBoundFieldCache()
 
+    def get_removed_fields(self, cleaned_data):
+        """
+        Build the list of fields to be removed due to conditional displays
+        """
+        # build a catalog of fields **targeted** by the conditions
+        condition_targets = {}
+
+        # For each condition, extract its status (should I display or not)
+        for condition in self._conditions:
+            # should we keep these fields?
+            keep_fields = condition.keep_fields(cleaned_data)
+            for field_id in condition.fields_ids:
+                # Fill the catalog
+                if field_id not in condition_targets:
+                    condition_targets[field_id] = []
+                condition_targets[field_id].append(keep_fields)
+        # Here, the catalog contains fields targeted by 1 or many conditions.
+
+        # If only one condition says "please display X", we'll keep X
+        # That's why we gather the conditions using "any"
+        condition_targets = {k: any(v) for k, v in condition_targets.items()}
+        # We'll only remove fields that are targeted by conditions **and**
+        # those conditions are false
+        to_remove = [k for k, v in condition_targets.items() if not v]
+        return to_remove
+
     def clean(self):
         cleaned_data = super(BaseDynamicForm, self).clean()
-        for condition in self._conditions:
-            cleaned_data = condition(self, cleaned_data)
+
+        removed_fields = self.get_removed_fields(cleaned_data)
```

Comment:
And I think that we can use a `generator` here instead of a `list`

Reply:
Fixed.